db/migrate/20180510184222_update_foreign_key_constraints.rb

```diff
@@ -0,0 +1,311 @@
+ALTER TABLE "traits" ADD CONSTRAINT "cultivar_exists" FOREIGN KEY ("cultivar_id") REFERENCES "cultivars" ("id") ON UPDATE CASCADE NOT VALID;
+ALTER TABLE "traits" ADD CONSTRAINT "treatment_exists" FOREIGN KEY ("treatment_id") REFERENCES "treatments" ("id") ON UPDATE CASCADE NOT VALID;
+ALTER TABLE "traits" ADD CONSTRAINT "variable_exists" FOREIGN KEY ("variable_id") REFERENCES "variables" ("id") ON UPDATE CASCADE NOT VALID;
+ALTER TABLE "traits" ADD CONSTRAINT "user_exists" FOREIGN KEY ("user_id") REFERENCES "users" ("id") ON UPDATE CASCADE NOT VALID;
+ALTER TABLE "traits" ADD CONSTRAINT "entity_exists" FOREIGN KEY ("entity_id") REFERENCES "entities" ("id") ON UPDATE CASCADE NOT VALID;
+ALTER TABLE "traits" ADD CONSTRAINT "method_exists" FOREIGN KEY ("method_id") REFERENCES "methods" ("id") ON UPDATE CASCADE NOT VALID;
+ALTER TABLE "treatments" ADD CONSTRAINT "user_exists" FOREIGN KEY ("user_id") REFERENCES "users" ("id") ON UPDATE CASCADE NOT VALID;
+ALTER TABLE "yields" ADD CONSTRAINT "site_exists" FOREIGN KEY ("site_id") REFERENCES "sites" ("id") ON UPDATE CASCADE NOT VALID;
+ALTER TABLE "yields" ADD CONSTRAINT "entity_exists" FOREIGN KEY ("entity_id") REFERENCES "entities" ("id") ON UPDATE CASCADE NOT VALID;
+}
+
+    ['Input', 'Model', 'Posterior'].each do |container_type|
+      execute %{
+    CREATE OR REPLACE FUNCTION forbid_dangling_#{container_type.downcase}_references() RETURNS TRIGGER AS $$
+BEGIN
+    IF
+        OLD.id = SOME(SELECT container_id FROM dbfiles WHERE container_type = '#{container_type}')
+        AND TG_OP = 'DELETE'
+    THEN
+        RAISE NOTICE 'You can''t remove the row with id % because it is referred to by some dbfile.', OLD.id;
+        RETURN NULL;
+    ELSIF
+        TG_OP = 'UPDATE'
+    THEN
+        RAISE NOTICE 'About to update container_id in rows of dbfiles table where container_type is #{container_type}.';
+        RAISE NOTICE 'For this to work, you should drop the "valid_#{container_type.downcase}_refs" constraint before updating #{container_type.downcase} ids and re-add it after you are done.';
+        UPDATE dbfiles SET container_id = NEW.id WHERE container_id = OLD.id AND container_type = '#{container_type}';
+        RETURN NEW;
+    END IF;
+
+    RETURN OLD;
+END;
+$$ LANGUAGE plpgsql;
+
+DROP TRIGGER IF EXISTS forbid_dangling_#{container_type.downcase}_references ON #{container_type.downcase}s;
+CREATE TRIGGER forbid_dangling_#{container_type.downcase}_references
+    BEFORE DELETE OR UPDATE OF id ON #{container_type.downcase}s
+FOR EACH ROW
+    EXECUTE PROCEDURE forbid_dangling_#{container_type.downcase}_references();
+
+}
+    end
+  end
+
+  def self.down
+	execute %q{
+ALTER TABLE "posteriors_ensembles" DROP CONSTRAINT "fk_posteriors_ensembles_ensembles_1";
+ALTER TABLE "posteriors_ensembles" ADD CONSTRAINT "fk_posteriors_ensembles_ensembles_1" FOREIGN KEY ("ensemble_id") REFERENCES "ensembles" ("id") ON DELETE CASCADE;
```

Comment:
why not on update cascade?

Reply:
Simply because that's how things are now—the down migration is meant to restore the database structure to its original state, that is, the state it was in before the migration was applied.